File: HuaweiVcLib/src/main/java/com/hw/huaweivclib/activity/CallerIDActivity.java
package com.hw.huaweivclib.activity;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.huawei.opensdk.callmgr.CallMgr;
import com.huawei.opensdk.ecsdk.common.UIConstants;
import com.hw.baselibrary.common.AppManager;
import com.hw.baselibrary.utils.sharedpreferences.SPStaticUtils;
import com.hw.huaweivclib.R;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * 来电显示界面
 */
public class CallerIDActivity extends BaseMediaActivity implements View.OnClickListener {
    private ImageView ivAvatar;
    private TextView tvNumber;
    private TextView tvHangUp;
    private TextView tvAnswer;

    @Override
    public void onClick(View v) {
        if (R.id.tv_hang_up == v.getId()) {
            hangUp();
        } else if (R.id.tv_answer == v.getId()) {
            answer();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    private boolean isFristClick = true;

    /**
     * 接听
     */
    private void answer() {
        if (isFristClick){
            CallMgr.getInstance().answerCall(mCallID, mIsVideoCall);
        }
    }

    /**
     * 挂断
     */
    private void hangUp() {
        //结束掉等待的对话框
        AppManager.Companion.getInstance().pushActivity(LoadingActivity.class);

        CallMgr.getInstance().endCall(mCallID);
        CallMgr.getInstance().stopPlayRingBackTone();
        CallMgr.getInstance().stopPlayRingingTone();
        finish();
    }

    @Override
    public void onBackPressed() {
//        super.onBackPressed();
//        hangUp();
    }

    @Override
    public int bindLayout() {
        return R.layout.activity_caller_id;
    }

    @Override
    public void initView(@Nullable Bundle savedInstanceState, @NotNull View contentView) {
        ivAvatar = (ImageView) findViewById(R.id.iv_avatar);
        tvNumber = (TextView) findViewById(R.id.tv_number);
        tvHangUp = (TextView) findViewById(R.id.tv_hang_up);
        tvAnswer = (TextView) findViewById(R.id.tv_answer);

        //是否移动端创建会议填false
        SPStaticUtils.put(UIConstants.IS_CREATE, false);
    }

    @Override
    public void doBusiness() {
        tvNumber.setText(TextUtils.isEmpty(String.valueOf(mCallNumber)) ? "" : String.valueOf(mCallNumber));
    }

    @Override
    public void setListeners() {
        tvHangUp.setOnClickListener(this);
        tvAnswer.setOnClickListener(this);
    }
}

File: vclibrary/src/main/java/com/zxwl/vclibrary/activity/CallerIDActivity.kt
package com.zxwl.vclibrary.activity

import android.view.View
import android.widget.ImageView
import android.widget.TextView
import com.huawei.opensdk.callmgr.CallMgr
import com.huawei.opensdk.commonservice.util.LogUtil
import com.huawei.opensdk.demoservice.MeetingMgr
import com.huawei.opensdk.esdk.utils.ActivityStack
import com.zxwl.vclibrary.R
import com.zxwl.vclibrary.util.NotificationUtils
import com.zxwl.vclibrary.util.ToastHelper
import com.zxwl.vclibrary.util.permission.PermissionUtils
import kotlinx.android.synthetic.main.activity_caller_id.*
import me.jessyan.autosize.internal.CancelAdapt

/**
 * 来电界面
 */
class CallerIDActivity : BaseMediaActivity(), View.OnClickListener,CancelAdapt {

    companion object {
        val TAG = CallerIDActivity::class.java.simpleName
    }

    override fun initData() {
        super.initData()

        tvNumber.text = mDisplayName

        if (mIsVideoCall) {
            tvCallContent.setText(if (0 == mConfToCallHandle) "想与您视频通话..." else "邀请您加入视频会议...")
            ivVoiceAnsewer.setVisibility(View.VISIBLE)
        } else {
            tvCallContent.setText(if (0 == mConfToCallHandle) "想与您音频通话..." else "邀请您加入语音会议...")
            ivVoiceAnsewer.setVisibility(View.GONE)
            tvAnswer.setCompoundDrawablesRelativeWithIntrinsicBounds(
                0,
                R.mipmap.ic_voice_conf_answer,
                0,
                0
            )
        }
    }

    override fun setListener() {
        super.setListener()
        tvHangUp.setOnClickListener(this)
        tvAnswer.setOnClickListener(this)
        ivVoiceAnsewer.setOnClickListener(this)
    }

    override fun getLayoutId(): Int = R.layout.activity_caller_id

    override fun onClick(v: View?) {
        when (v) {
            tvHangUp -> {
                hangUp()
            }

            tvAnswer -> {
                answer(mIsVideoCall)
            }

            ivVoiceAnsewer -> {
                answer(false)
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        NotificationUtils.cancel(NotificationUtils.CALL_IN_ID)
    }

    /**
     * 接听
     */
    private fun answer(isVideoCall: Boolean) {
        PermissionUtils.requestCameraAndMicroPermission(object : PermissionUtils.FullCallback {
            override fun onGranted(permissionsGranted: List<String>) {
                //会议
                if (callInfo!!.isFocus) {
                    CallMgr.getInstance().stopPlayRingingTone()
                    CallMgr.getInstance().stopPlayRingBackTone()

                    LogUtil.i(TAG, "走到接听的回调中来，isVideoCall：$isVideoCall")

                    CallMgr.getInstance().answerCall(mCallID, isVideoCall)
                    MeetingMgr.getInstance().acceptConf(isVideoCall)
                    finish()
                } else {//视频呼叫
                    CallMgr.getInstance().answerCall(mCallID, isVideoCall)
                }
            }

            override fun onDenied(
                permissionsDeniedForever: List<String>,
                permissionsDenied: List<String>
            ) {
                if (isVideoCall) {
                    ToastHelper.showShort("接听需要获取摄像头和麦克风权限")
                } else {
                    ToastHelper.showShort("接听需要获取麦克风权限")
                }
                hangUp()
            }
        })
    }

    /**
     * 挂断
     */
    private fun hangUp() {
        //结束掉等待的对话框
        ActivityStack.getIns().popup(LoadingActivity::class.java)

        if (0 == mConfToCallHandle) {
            CallMgr.getInstance().endCall(mCallID)
        } else {
            CallMgr.getInstance().stopPlayRingingTone()
            CallMgr.getInstance().stopPlayRingBackTone()

            CallMgr.getInstance().endCall(mCallID)
            MeetingMgr.getInstance().rejectConf()
        }
        finish()
    }

    override fun onBackPressed() {

    }

}